File: tests/algorithms/x/Pascal/physics/centripetal_force.pas
{$mode objfpc}{$modeswitch nestedprocvars}
program Main;
uses SysUtils, Math;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
procedure error(msg: string);
begin
  panic(msg);
end;
function _floordiv(a, b: int64): int64; var r: int64;
begin
  r := a div b;
  if ((a < 0) xor (b < 0)) and ((a mod b) <> 0) then r := r - 1;
  _floordiv := r;
end;
function _to_float(x: int64): real;
begin
  _to_float := x;
end;
function to_float(x: int64): real;
begin
  to_float := _to_float(x);
end;
procedure json(xs: array of real); overload;
var i: integer;
begin
  write('[');
  for i := 0 to High(xs) do begin
    write(xs[i]);
    if i < High(xs) then write(', ');
  end;
  writeln(']');
end;
procedure json(x: int64); overload;
begin
  writeln(x);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
function centripetal(centripetal_mass: real; centripetal_velocity: real; centripetal_radius: real): real; forward;
function floor(floor_x: real): real; forward;
function pow10(pow10_n: int64): real; forward;
function round(round_x: real; round_n: int64): real; forward;
procedure show(show_mass: real; show_velocity: real; show_radius: real); forward;
function centripetal(centripetal_mass: real; centripetal_velocity: real; centripetal_radius: real): real;
begin
  if centripetal_mass < 0 then begin
  panic('The mass of the body cannot be negative');
end;
  if centripetal_radius <= 0 then begin
  panic('The radius is always a positive non zero integer');
end;
  exit(((centripetal_mass * centripetal_velocity) * centripetal_velocity) / centripetal_radius);
end;
function floor(floor_x: real): real;
var
  floor_i: integer;
begin
  floor_i := Trunc(floor_x);
  if Double(floor_i) > floor_x then begin
  floor_i := floor_i - 1;
end;
  exit(Double(floor_i));
end;
function pow10(pow10_n: int64): real;
var
  pow10_p: real;
  pow10_i: int64;
begin
  pow10_p := 1;
  pow10_i := 0;
  while pow10_i < pow10_n do begin
  pow10_p := pow10_p * 10;
  pow10_i := pow10_i + 1;
end;
  exit(pow10_p);
end;
function round(round_x: real; round_n: int64): real;
var
  round_m: real;
begin
  round_m := pow10(round_n);
  exit(Floor((round_x * round_m) + 0.5) / round_m);
end;
procedure show(show_mass: real; show_velocity: real; show_radius: real);
var
  show_f: real;
begin
  show_f := centripetal(show_mass, show_velocity, show_radius);
  writeln(FloatToStr(round(show_f, 2)));
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  show(15.5, -30, 10);
  show(10, 15, 5);
  show(20, -50, 15);
  show(12.25, 40, 25);
  show(50, 100, 50);
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
  writeln('');
end.

File: tests/algorithms/x/Clojure/physics/centripetal_force.clj
(ns main (:refer-clojure :exclude [centripetal floor pow10 round show]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(defn toi [s]
  (int (Double/valueOf (str s))))

(defn _ord [s]
  (int (first s)))

(defn mochi_str [v]
  (cond (float? v) (let [s (str v)] (if (clojure.string/ends-with? s ".0") (subs s 0 (- (count s) 2)) s)) :else (str v)))

(defn _fetch [url]
  {:data [{:from "" :intensity {:actual 0 :forecast 0 :index ""} :to ""}]})

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare centripetal floor pow10 round show)

(declare _read_file)

(def ^:dynamic floor_i nil)

(def ^:dynamic pow10_i nil)

(def ^:dynamic pow10_p nil)

(def ^:dynamic round_m nil)

(def ^:dynamic show_f nil)

(defn centripetal [centripetal_mass centripetal_velocity centripetal_radius]
  (try (do (when (< centripetal_mass 0.0) (throw (Exception. "The mass of the body cannot be negative"))) (when (<= centripetal_radius 0.0) (throw (Exception. "The radius is always a positive non zero integer"))) (throw (ex-info "return" {:v (/ (*' (*' centripetal_mass centripetal_velocity) centripetal_velocity) centripetal_radius)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn floor [floor_x]
  (binding [floor_i nil] (try (do (set! floor_i (long floor_x)) (when (> (double floor_i) floor_x) (set! floor_i (- floor_i 1))) (throw (ex-info "return" {:v (double floor_i)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn pow10 [pow10_n]
  (binding [pow10_i nil pow10_p nil] (try (do (set! pow10_p 1.0) (set! pow10_i 0) (while (< pow10_i pow10_n) (do (set! pow10_p (*' pow10_p 10.0)) (set! pow10_i (+' pow10_i 1)))) (throw (ex-info "return" {:v pow10_p}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn round [round_x round_n]
  (binding [round_m nil] (try (do (set! round_m (pow10 round_n)) (throw (ex-info "return" {:v (/ (floor (+' (*' round_x round_m) 0.5)) round_m)}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn show [show_mass show_velocity show_radius]
  (binding [show_f nil] (do (set! show_f (centripetal show_mass show_velocity show_radius)) (println (mochi_str (round show_f 2))) show_mass)))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (show 15.5 (- 30.0) 10.0)
      (show 10.0 15.0 5.0)
      (show 20.0 (- 50.0) 15.0)
      (show 12.25 40.0 25.0)
      (show 50.0 100.0 50.0)
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)
